### ISSUE TITLE
HKDF context re-use issue

### ISSUE THREAD

**Issue by dlegaultbbry** [openssl/openssl#7497]
Normally with all openssl algorithms I've used (at least with cipher, digest and a few others), if you re-init the context structure, it can be re-used to perform another round of the same algorithm.

But for some reason this doesn't work for HKDF.

This demo code was compiled and linked with openssl 1.1.1 on Linux

```
#include <string.h>
#include <stdio.h>

#include <openssl/evp.h>
#include <openssl/kdf.h>

int main(int argc, char *argv[])
{
	int i;
	int try = 0;
	EVP_PKEY_CTX	*pctx;
	const EVP_MD	*md;

	uint8_t inkey[] =	"\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b";
	uint8_t salt[] =	"\x00\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a\x0b\x0c";
	uint8_t info[] =	"\xf0\xf1\xf2\xf3\xf4\xf5\xf6\xf7\xf8\xf9";
	uint8_t outkey[] =	"\x08\x5a\x01\xea\x1b\x10\xf3\x69\x33\x06\x8b\x56\xef\xa5\xad\x81\xa4\xf1\x4b\x82\x2f\x5b\x09\x15\x68\xa9\xcd\xd4\xf1\x55\xfd\xa2\xc2\x2e\x42\x24\x78\xd3\x05\xf3\xf8\x96";
	uint8_t outkey_cmp[sizeof(outkey)-1];
	size_t keysize = sizeof(outkey)-1;

	md = EVP_get_digestbyname("sha1");

	pctx = EVP_PKEY_CTX_new_id(EVP_PKEY_HKDF, NULL);

restart:

	if (EVP_PKEY_derive_init(pctx) <= 0) {
		printf("failed init HKDF key\n");
		goto done;
	}

	if (EVP_PKEY_CTX_set_hkdf_md(pctx, md) <= 0) {
		printf("failed to set HKDF digest\n");
		goto done;
	}

	if (EVP_PKEY_CTX_set1_hkdf_salt(pctx, salt, sizeof(salt)-1) <= 0) {
		printf("failed to set HKDF salt\n");
		goto done;
	}

	if (EVP_PKEY_CTX_set1_hkdf_key(pctx, inkey, sizeof(inkey)-1) <= 0) {
		printf("failed to set HKDF secret\n");
		goto done;
	}

	if (EVP_PKEY_CTX_add1_hkdf_info(pctx, info, sizeof(info)-1) <= 0) {
		printf("failed to set HKDF info\n");
		goto done;
	}

	if (EVP_PKEY_derive(pctx, outkey_cmp, &keysize) <= 0) {
		printf("failed to derive HKDF key\n");
		goto done;
	}

	printf("orig key: ");
	for (i=0;i<keysize;i++) {
		printf("%.2x", outkey[i]);
	}
	puts("\n");
	printf(" kdf key: ");
	for (i=0;i<keysize;i++) {
		printf("%.2x", outkey_cmp[i]);
	}
	puts("\n");

	if (memcmp(outkey, outkey_cmp, keysize)) {

		printf("keys don't match!\n");
	} else {
		printf("keys match!\n");
	}

	if (try < 4) {
		try++;
		goto restart;
	}

done:

	return 0;
}
```

**Comment by mattcaswell** [comment]
Hmm, `EVP_PKEY_derive_init` doesn't seem to reset *any* of the previous parameters. In this case this means that the `EVP_PKEY_CTX_add1_hkdf_info()` call appends info to the previous info each time around the loop - rather than starting again.

**Comment by mattcaswell** [comment]
Yes - I think it should. Patch on the way.

**Comment by mattcaswell** [comment]
Fix in #7501.
